edgeless_node/src/agent/mod.rs

```diff
@@ -65,33 +67,110 @@ impl Agent {
 
+        // After spawning a new function, the function´s class is only used to determine which runner to start it on.
+        // When stopping, only the stop_function_id is provided which does not allow to know which runner it is
+        // currently deployed on. Here, we implement a instance_id -> function_class HashMap
+        let mut component_id_to_class_map = std::collections::HashMap::<edgeless_api::function_instance::ComponentId, String>::new();
+
         log::info!("Starting Edgeless Agent");
         while let Some(req) = receiver.next().await {
             match req {
                 AgentRequest::Spawn(spawn_req) => {
                     log::debug!("Agent Spawn {:?}", spawn_req);
-                    match runner.start(spawn_req).await {
-                        Ok(_) => {}
-                        Err(err) => {
-                            log::error!("Unhandled Start Error: {}", err);
+
+                    // Save function_class for further interaction.
+                    // We can assume that the Optional<instance_id> is present.
+                    if spawn_req.instance_id.is_none() {
+                        log::error!("No instance_id provided for SpawnFunctionRequest!");
+                        continue;
+                    }
+                    component_id_to_class_map.insert(
+                        spawn_req.instance_id.clone().unwrap().function_id,
+                        spawn_req.code.function_class_type.clone(),
+                    );
+
+                    // Get runner for function_class of spawn_req
+                    match runners.get_mut(&spawn_req.code.function_class_type) {
+                        Some(r) => {
+                            // Forward the start request to the correct runner
+                            match r.start(spawn_req).await {
+                                Ok(_) => {}
+                                Err(err) => {
+                                    log::error!("Unhandled Start Error: {}", err);
+                                    continue;
+                                }
+                            }
+                        }
+                        None => {
+                            log::warn!("Could not find runner for {}", spawn_req.code.function_class_type);
+                            continue;
                         }
                     }
                 }
                 AgentRequest::Stop(stop_function_id) => {
                     log::debug!("Agent Stop {:?}", stop_function_id);
-                    match runner.stop(stop_function_id).await {
-                        Ok(_) => {}
-                        Err(err) => {
-                            log::error!("Unhandled Stop Error: {}", err);
+
+                    // Get function class by looking it up in the instanceId->functionClass map
+                    let function_class: String;
```

Comment:
Rust allows you to use the match as an expression, so you can assign this directly from the match.

Reply:
done in both cases